### PR TITLE
One operator class wrapping generated functions

### PR DESCRIPTION
**Pull request by makaimann**
Finished the updates to functions I mentioned in our meeting today and also incorporated your comments on how to handle them.

I verified that does this not break the Travis tests in the Stanford AHA project (although some of that code can be simplified because of these changes).

Personally I like this structure a lot. Primarily because it makes adding functions very simple. But please let me know if you think there's a better way or if the operator/generators are too confusing. My hope is that users would never have to interact with those at all. They would just add function names/parameters to a dict and the rest happens magically. See below for change summary and small example.

Summary of changes:

There is a class, operator, which uses the type functools.partial to hold partial evaluations of functions.

Given a name and a namedtuple, fdata, functions are automatically generated.
fdata contains the number of indices (this is for indexed operators, it's 0 for non-indexed operators), the minimum arity, and the maximum arity.

Non-indexed operators are immediately wrapped by the operator class. Thus, they look like operators with zero index arguments. Indexed operators are a closure which calls generate_operator. This allows you to either evaluate the function entirely or only add the index arguments. However, you cannot pass less than the number of index arguments, or any number of arguments between num_indexes and min_arity.

i.e.

functions.bvult(bv1) <--- not allowed
functions.extract(4) <--- not allowed

functions.bvult(bv1, bv2) <--- allowed
functions.extract(4, 2) <--- allowed
functions.extract(4, 2, bv1) <--- allowed

Small example:

from smt_switch import functions, solvers, sorts

\# prints <operator: bvult, () {}>
\# i.e. it is an operator named bvult with no positional arguments and no keyword arguments
print(functions.bvult)

s = solvers.BoolectorSolver()
bv = s.declare_const('bv', sorts.BitVec(8))

ex4_2 = functions.extract(4, 2)

\# the following are all equivalent
bvex1 = ex4_2(bv)
bvex2 = s.apply_fun(ex4_2, bv)
bvex3 = functions.extract(4, 2, bv)  \# syntactic sugar for s.apply_fun(ex4_2, bv)

s.Assert(bvex1 == bvex2 == bvex3)  \# trivially true

assert s.check_sat()

### CODE REVIEW

File: smt_switch/src/functions.py [hunk 1]
```diff
@@ -1,480 +1,222 @@
...
+         'extract': fdata(2, 1, 1), 'concat': fdata(0, 2, 2), 'zero_extend': fdata(0, 2, 2),
+         'bvand': fdata(0, 2, 2), 'bvor': fdata(0, 2, 2), 'bvxor': fdata(0, 2, 2),
+         'bvadd': fdata(0, 2, 2), 'bvsub': fdata(0, 2, 2), 'bvmul': fdata(0, 2, 2),
+         'bvudiv': fdata(0, 2, 2), 'bvurem': fdata(0, 2, 2), 'bvshl': fdata(0, 2, 2),
+         'bvashr': fdata(0, 2, 2), 'bvlshr': fdata(0, 2, 2), 'bvult': fdata(0, 2, 2),
+         'bvule': fdata(0, 2, 2), 'bvugt': fdata(0, 2, 2), 'bvuge': fdata(0, 2, 2),
+         'bvslt': fdata(0, 2, 2), 'bvsle': fdata(0, 2, 2), 'bvsgt': fdata(0, 2, 2),
+         'bvsge': fdata(0, 2, 2), 'bvnot': fdata(0, 1, 1), 'bvneg': fdata(0, 1, 1),
+         'No_op': fdata(0, 0, 0), }
+
```
Comment (mpreiner) [inline smt_switch/src/functions.py]: Use consistent function names (uppercase vs. lowercase). Bvudiv, ...

File: smt_switch/src/functions.py [hunk 2]
```diff
@@ -1,124 +1,222 @@
...
+                                 ' of type [Solver Name]Term')
+
+            if config.strict and len(args) > fdata.max_arity:
+                raise ValueError('In strict mode and received {} args when max arity = '
+                                 .format(len(args), fdata.max_arity))
+
             return s_terms[0].solver.apply_fun(fun, *args, **kwargs)
 
-        # execution should never reach this code
-        assert True == False, 'An invariant is broken'
```
Comment (aniemetz) [inline smt_switch/src/functions.py]: This is only minor, but `assert False, '...' ` would suffice ;-)

File: smt_switch/src/solvers/BoolectorSolver.py [hunk 3]
```diff
@@ -92,25 +93,26 @@ def set_nonstandard_option(self, optionstr, value):
     def declare_const(self, name, sort):
```
Comment (mpreiner) [inline smt_switch/src/solvers/BoolectorSolver.py]: In SMT-LIB v2.6 declare-const is used to define nullary function symbols. What about function symbols with arity > 0?

File: smt_switch/src/functions.py [hunk 4]
```diff
@@ -1,480 +1,222 @@
...
-             'max': 2}
+__all__ = ['And', 'Or']  # append to list in generate functions
 
-    def __init__(self):
-        super().__init__(self.arity, '(> arg1 arg2)')
 
-    def osort(self, *args):
-        return sorts.Bool()
+fdata = namedtuple('fdata', 'num_indices, min_arity, max_arity')
+__MAXARGS__ = 6000
```
Comment (mpreiner) [inline smt_switch/src/functions.py]: Is there a reason why 6000? Would it make sense to use -1 for an arbitrary number of arguments?

**Comment by aniemetz** [comment]
Currently the SMT-LIB commands are implemented for each solver independently. Which entails that you always have to keep track of SMT-LIB commands in multiple places. Wouldn't it make sense to have one SMT-LIB API object, e.g., `smtlibapi = SMTLibAPI (Boolector())`, and this object is initialized with a solver object, and then you can say, e.g., `smtlibapi.check_sat()` and the API object knows which solver to pass the check_sat command to? This would have the advantage that you have one API class that implements all the SMT-LIB commands (easy to keep track of), and you hook in the solver-specific command implementations on initialization.

File: smt_switch/src/functions.py (hunk 2 above)
Comment (makaimann) [inline smt_switch/src/functions.py]: Ope, duhh. Silly me. This is gone in the latest commit anyway.

File: smt_switch/src/solvers/BoolectorSolver.py (hunk 3 above)
Comment (makaimann) [inline smt_switch/src/solvers/BoolectorSolver.py]: Unfortunately I don't have a declare-fun equivalent yet, that needs to be added still.

File: smt_switch/src/functions.py (hunk 4 above)
Comment (makaimann) [inline smt_switch/src/functions.py]: I'll have to take a look at this again. At one point I thought I was having an issue unpacking tuples with * that were too large, but I think that might have turned out to be unrelated. In which case -1 would be much better.

File: smt_switch/src/solvers/BoolectorSolver.py (hunk 3 above)
Comment (mpreiner) [inline smt_switch/src/solvers/BoolectorSolver.py]: declare-const can then be a special (nullary) case of declare-fun

**Pull request merged**